### PR TITLE
Solve leetcode problem `997 - Find the Town Judge`

### PR DESCRIPTION
**Pull request by joshiayush**
Add function to find that one single person in the town given as form of an array that
does not trust anyone around the town but everyone trusts this person

Signed-off-by: Ayush Joshi <ayush854032@gmail.com>

### CODE REVIEW

**Review by anthonygraca — CHANGES_REQUESTED** [review]
Really good job following the general style of the project and finding the optimal solution.


Here are a few issues I would like you to consider:

1.) One of the constraints state that n could only be a value between 1 and 1000. `int64_t` may not be preferable compared to `uint16_t` since the input parameters would not make use of most of the bits in an `int64_t`. Consider refactoring code to use of `uint16_t`. I am open to the idea of just using an `int`. Let me know if this is a change you are willing to make.

2.)  Could you explain why the time complexity is O(2^n / 2)? I think the time complexity is O(N + T), where N = n and T = trust.size()

3.) Good job in using `const` when you can. This good practice is encouraged.

4.) It is a code smell when there is a new operator but there is not a delete operator at the end of the function as that leads to memory leaks. Consider using a vector with `std::vector<int> countTrustArray{n+1, 0};` This automatically creates a vector of size n+1 with a default value of 0 so the for-loop on line 48 & 49 isn't needed. I recommend making this change because it cuts down on a decent amount of code.

5.) nit. Usually it isn't recommended to name a variable with the underlying data structure. For example, "studentNameString" is better named as "studentName". This same idea applies to "countTrustArray". This isn't a big deal but conceptually, the whole point of this array is to track the indegree and outdegree of a directed graph and I feel that the idea doesn't show with the name of the data structure. I don't have a good alternative name and I think "countTrustArray" is fine given that variable names do not matter in an interview setting.

**Comment by joshiayush** [comment]
> 1.) One of the constraints state that n could only be a value between 1 and 1000. `int64_t` may not be preferable compared to `uint16_t` since the input parameters would not make use of most of the bits in an `int64_t`. Consider refactoring code to use of `uint16_t`. I am open to the idea of just using an int. Let me know if this is a change you are willing to make.

@anthonygraca Sure, I'm willing to change it the reason I put `int64_t` there is because after reading a lot of C/C++ code I found out that it is a common practice to use `64-bit` of type to represent buffer length, I'm willing to be corrected here if I'm wrong.

**Comment by joshiayush** [comment]
@anthonygraca Thanks for this `std::vector<int> countTrustArray{n+1, 0};` I'm gonna update my code shortly.

**Comment by joshiayush** [comment]
> 2.) Could you explain why the time complexity is O(2^n / 2)? I think the time complexity is O(N + T), where N = n and T = trust.size()

@anthonygraca Thanks for pointing this mistake out I was actually trying to formulate the `trust` size based on `n` and at that time I thought that the `trust` size totally depends on the number of people in that town in worst case it would be (if we have 3 people),

```
[[1, 2], [1, 3], [2, 1], [2, 3], [3, 1], [3, 2]]
```

But now I totally lost the picture I had last day about this and I don't know how to formulate it, please help.

**Comment by anthonygraca** [comment]
> @anthonygraca Sure, I'm willing to change it the reason I put int64_t there is because after reading a lot of C/C++ code I found out that it is a common practice to use 64-bit of type to represent buffer length, I'm willing to be corrected here if I'm wrong.

So the main reason why we use standard types (int8_t, int16_t, int32_t, int64_t, uint8_t...) is because although the C & C++ standard guarantees the size of a char to be 8 bits, the size of a short is not guaranteed to be 16 bits and the size of an int is not guaranteed to be 32 bits. The size of a general int data type changes depending on the implementation of the compiler and architecture. What this means is that an `int64_t` is used to guarantee the integer value will use 64 bits.

To be completely honest. It's not a big deal to use fixed-width types, especially during an interview. I think it would be better to just code how you would in an interview

**Comment by anthonygraca** [comment]
> @anthonygraca Thanks for pointing this mistake out I was actually trying to formulate the trust size based on n and at that time I thought that the trust size totally depends on the number of people in that town in worst case it would be (if we have 3 people),

The worst case is trust.size() == 1000

**Comment by joshiayush** [comment]
> So the main reason why we use standard types (int8_t, int16_t, int32_t, int64_t, uint8_t...) is because although the C & C++ standard guarantees the size of a char to be 8 bits, the size of a short is not guaranteed to be 16 bits and the size of an int is not guaranteed to be 32 bits. The size of a general int data type changes depending on the implementation of the compiler and architecture. What this means is that an int64_t is used to guarantee the integer value will use 64 bits.

@anthonygraca Thank you.

**Comment by joshiayush** [comment]
> The worst case is trust.size() == 1000

@anthonygraca Are you sure I should go like this and not formulate the number of people who've got the trust of others and who trusts others in that town based on the number of people?

**Comment by anthonygraca** [comment]
> But now I totally lost the picture I had last day about this and I don't know how to formulate it, please help.

If you are asking how it is O(N + T). The idea is that you have two for loops. Your first loop iterates over the length of the trust vector and your second loop iterates over the n. This means your algorithm takes time proportional to O(length of trust vector + n). If we just let T = length of trust vector and N = n, then we can just say the worst case is O(T + N). 

Did that answer your question?

**Comment by anthonygraca** [comment]
> > The worst case is trust.size() == 1000
> 
> @anthonygraca Are you sure I should go like this and not formulate the number of people who've got the trust of others and who trusts others in that town based on the number of people?

Your algorithm does that already on the first for-loop. Leetcode says under the constraints that trust.size() will only be a maximum of 1000 elements.

**Comment by joshiayush** [comment]
> T = length of trust vector and N = n

@anthonygraca Before I used to formulate the time complexity based on one single value given to me and this time I was doing it around `n` but now you've cleared my doubt that I can always use other values not always formulating around one single value works. Thanks.

**Comment by anthonygraca** [comment]
> > T = length of trust vector and N = n
> 
> @anthonygraca Before I used to formulate the time complexity based on one single value given to me and this time I was doing it around `n` but now you've cleared my doubt that I can always use other values not always formulating around one single value works. Thanks.

I understand what you mean. Yes, the time/space complexity will not always be some version of "O(n), O(logn), etc". Time/Space can have multiple variables. LC #4 is another example of two variables affecting the time complexity.

File: src/main/cpp/easy/997-find-the-town-judge.h [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+ public:
+  // Finds the one single person that is the town judge.
+  //
+  // This function finds the one single person that does not trust anybody in
+  // the town but everybody trusts him/her and this person is basically judging
+  // every person around the town; to make sure that you don't mess up in front
+  // of him/her we need to find this person.
+  ::std::int64_t findJudge(
+      ::std::uint16_t n,
+      const ::std::vector<::std::vector<::std::int64_t>>& trust) {
```
Comment (joshiayush) [inline src/main/cpp/easy/997-find-the-town-judge.h:41]: Trust size is always going to be a lot larger than the number of people we have in that town, so I don't feel like I should change `int64_t` to `uint16_t` although I can change it to `uint64_t` but keeping it just `int64_t` is not gonna have a big difference. What do you say?

**Comment by joshiayush** [comment]
> LC #4 is another example of two variables affecting the time complexity.

Yes I've done that 😂.

File: src/main/cpp/easy/997-find-the-town-judge.h (hunk 1 above)
Comment (anthonygraca) [reply src/main/cpp/easy/997-find-the-town-judge.h:41]: Leetcode states that the constraint on trust.size() is max 1000 elements. So I think the int64_t is overkill. But you are right, functionally the code still works.

Comment (joshiayush) [reply src/main/cpp/easy/997-find-the-town-judge.h:41]: Anything else you would like me to change?

Comment (anthonygraca) [reply src/main/cpp/easy/997-find-the-town-judge.h:41]: For those lines specifically, prefer `std::uint16_t` over `::std::uint16_t`. [This stackoverflow post explains the difference in the syntax](https://stackoverflow.com/questions/33102207/difference-in-using-namespace-std-vs-std). I don't plan for this project to shadow the `std` namespace.

**Comment by anthonygraca** [comment]
Looks good. Good job in resolving that memory leak. I feel comfortable merging this in. 

For style, prefer [the Google Style Guide](https://google.github.io/styleguide/cppguide.html). I am not 100% compliant but I will be following it more closely for future problems. 

Great job Ayush!

**Comment by anthonygraca** [comment]
*Reopening to merge in

**Comment by joshiayush** [comment]
> For style, prefer the Google Style Guide.

@anthonygraca Mmm... I always follow Google Style why are you saying this? have a look into my source code in my repositories you'll find Google Style there and I'm making these PRs in that style too... I don't know 😂 why you said that or I didn't understand you I guess.

**Pull request merged**